x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeValues.java

```diff
@@ -0,0 +1,496 @@
+    }
+
+    private final NewLeafLoader newLeafLoader;
+
+    public StringRuntimeValues(NewLeafLoader newLeafLoader) {
+        this.newLeafLoader = newLeafLoader;
+    }
+
+    public CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+        return unstarted().docValues();
+    }
+
+    public Query existsQuery(String fieldName) {
+        return unstarted().new ExistsQuery(fieldName);
+    }
+
+    public Query fuzzyQuery(String fieldName, String value, int maxEdits, int prefixLength, int maxExpansions, boolean transpositions) {
+        return unstarted().new FuzzyQuery(fieldName, value, maxEdits, prefixLength, maxExpansions, transpositions);
+    }
+
+    public Query prefixQuery(String fieldName, String value) {
+        return unstarted().new PrefixQuery(fieldName, value);
+    }
+
+    public Query rangeQuery(String fieldName, String lowerValue, String upperValue, boolean includeLower, boolean includeUpper) {
+        return unstarted().new RangeQuery(fieldName, lowerValue, upperValue, includeLower, includeUpper);
+    }
+
+    public Query regexpQuery(String fieldName, String pattern, int flags, int maxDeterminizedStates) {
+        return unstarted().new RegexpQuery(fieldName, pattern, flags, maxDeterminizedStates);
+    }
+
+    public Query termQuery(String fieldName, String value) {
+        return unstarted().new TermQuery(fieldName, value);
+    }
+
+    public Query termsQuery(String fieldName, String... values) {
+        return unstarted().new TermsQuery(fieldName, values);
+    }
+
+    public Query wildcardQuery(String fieldName, String pattern) {
+        return unstarted().new WildcardQuery(fieldName, pattern);
+    }
+
+    @Override
+    protected SharedValues newSharedValues() {
+        return new SharedValues();
+    }
+
+    protected class SharedValues extends AbstractRuntimeValues<SharedValues>.SharedValues {
+        private String[] values = new String[1];
+
+        @Override
+        protected IntConsumer newLeafLoader(LeafReaderContext ctx) throws IOException {
+            return newLeafLoader.leafLoader(ctx, this::add);
+        }
+
+        private void add(String value) {
+            if (value == null) {
+                return;
+            }
+            int newCount = count + 1;
+            if (values.length < newCount) {
+                values = Arrays.copyOf(values, ArrayUtil.oversize(newCount, RamUsageEstimator.NUM_BYTES_OBJECT_REF));
+            }
+            values[count] = value;
+            count = newCount;
+        }
+
+        @Override
+        protected void sort() {
+            Arrays.sort(values, 0, count);
+        }
+
+        private CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+            alwaysSortResults();
+            return DocValues::new;
+        }
+
+        private class DocValues extends SortedBinaryDocValues {
```

Comment:
this is effectively the docvalues implementation that replaces the one that I had written which you deleted?

I was wondering, and not even sure if it matters, queries don't need sorted doc_values right? But we always implement the interface that we need for fielddata to simplify things? I know that I had implemented the lucene one for query but it made things more complex for almost no reason, I think.

Reply:
> this is effectively the docvalues implementation that replaces the one that I had written which you deleted?

Yeah. I had mine sitting around and it had unit tests and stuff. Its the same deal.

> I was wondering, and not even sure if it matters, queries don't need sorted doc_values right? But we always implement the interface that we need for fielddata to simplify things? I know that I had implemented the lucene one for query but it made things more complex for almost no reason, I think.

In the patch as it stands now the queries don't use this and look at the array of values directly. And they don't sort them unless they build this thing. When I was writing the queries it felt simpler not to target the doc values implementation - partially because most queries don't care about the values being sorted. And, if they *did* care about the values being sorted, then they'd want a sorted array. Which is what we use to build the doc values implementation anyway.